spring-ai-core/src/main/java/org/springframework/ai/chat/model/AbstractToolCallSupport.java

```diff
@@ -240,10 +241,11 @@ protected boolean isToolCall(ChatResponse chatResponse, Set<String> toolCallFini
 	protected boolean isToolCall(Generation generation, Set<String> toolCallFinishReasons) {
 		var finishReason = (generation.getMetadata().getFinishReason() != null)
 				? generation.getMetadata().getFinishReason() : "";
-		return generation.getOutput().hasToolCalls() && toolCallFinishReasons.stream()
+		ChatGenerationMetadata metadata = generation.getMetadata();
+		return generation.getOutput().hasToolCalls() && (toolCallFinishReasons.stream()
 			.map(s -> s.toLowerCase())
 			.toList()
-			.contains(finishReason.toLowerCase());
+			.contains(finishReason.toLowerCase()) || metadata != null);
```

Comment:
> > in the metadata but by the presence or absence of usage field.
> 
> This won't work, as `ChatGenerationMetadata` does not contain usage and it is quite likely always not null. The Usage field in in ChatResponseMetadata.
> 
> I can't tell from the docs on this model, but if they are trying to support an OpenAI API, then they should really implement finish_reason, it is common across all AI models we have, 12 of which extend this abstract base class.

Thanks for your reply. Yes I found the problem and I will fix it.

Reply:
The `.contains(finishReason.toLowerCase()) || metadata != null);` or event version with Generation`usage` is likely not going to be good condition as it can break many of the other implementations.

Reply:
> The `.contains(finishReason.toLowerCase()) || metadata != null);` or event version with Generation`usage` is likely not going to be good condition as it can break many of the other implementations.

Is there any good method? I can only think of this way, or should I add a new model for xinghuo separately?

Reply:
PTAL @tzolov @markpollack

Reply:
Is this model trying to support an OpenAI API?

Not having finish reason seems a design mistake in their API.  I want to limit the number of new models as much as possible and it seems like this one is close to openai api compatability.  If there is a method that can be made protected in the current OpenAiChatModel or some other seam introduced that would a good way to approach the problem by adding only a small amount of code to support this model.  I don't want to have a new model introduced that cut-n-pastes 99% of the current OpenAIChatModel as it will be maintance issue - and we actually already have this problem and don't want to make the problem worse.

Reply:
> Is this model trying to support an OpenAI API?
> 
> Not having finish reason seems a design mistake in their API. I want to limit the number of new models as much as possible and it seems like this one is close to openai api compatability. If there is a method that can be made protected in the current OpenAiChatModel or some other seam introduced that would a good way to approach the problem by adding only a small amount of code to support this model. I don't want to have a new model introduced that cut-n-pastes 99% of the current OpenAIChatModel as it will be maintance issue - and we actually already have this problem and don't want to make the problem worse.

I agree with your idea. Maybe adding a `modelType` parameter in the openai module can solve this problem, and i can rewrite some methods through modelType=xinghuo?